### PR TITLE
Multiple object store functionality

### LINKED ISSUE #172: ObjectStore integrity check
We could do with an object store integrity checker to ensure that when data is deleted everything is then saved correctly and all references to deleted UUIDs removed.

### LINKED ISSUE #319: Rework `assign_data` function to avoid metadata hoop jumping
At the moment `assign_data` expects a dictionary of metadata that's extracted from a data dictionary, this is really convoluted as we have to extract the data and then pass it in.

### LINKED ISSUE #410: Improve metadata  / metastore handling
At the moment we load in the metastore of each object, use a function to read it and then have to close it manually.
Could we create a thin wrapper / context manager to handle the loading and searching of each object's metastore?

### LINKED ISSUE #496: Allow users to access data from a group object store and a local one
We need to allow users to be able to access data from multiple object stores at once - their own local one and a group object store. This also means that it needs to be easy to change and work out which object store is being written to when adding data.

This may supercede Issue #262.

### PR DESCRIPTION
**Pull request by gareth-j**
This adds multiple object store functionality to OpenGHG by
- Updating the config file to handle multiple object stores and adds permissions
- Updates `openghg --quickstart` through `openghg.util.create_config`
- Adds a context manager to the store classes `ObsSurface` etc

From discussion on https://github.com/openghg/openghg/issues/496

- [x] - Multiple object stores should be treated equally (no "precedence" between them)
- [x] - Add ability to specify which object store to add to, search and to retrieve from
- [x]    - Always have to specify which one to add to for now? - No need to specify if only a single writable store
- [x]    - But don't need to explicitly specify which object store to search / retrieve from? Search defaults to searching all stores
- [x] - Allow searching all object stores (can return multiple SearchResults objects)
- [x] - Allow retrieving from all object stores, as long as there is no ambiguity otherwise fail
- [x] - Allow read/write equally from/to all object stores for now (consider read-only object stores in future or allow this be managed through group assignment and permissions)

- [x]  Integrity checker
- [x] Update tutorials
- [x] Add to changelog

- ~~zarr / better compression of data in store~~ - will do in separate PR

Closes #319
Closes #410 
Closes #172 
Closes #496

### CODE REVIEW

File: doc/source/tutorials/local/Adding_new_data/Select_write_store.rst
Comment (rt17603) [inline doc/source/tutorials/local/Adding_new_data/Select_write_store.rst:41]: Does this allow more than 2 object stores to be added? Would be nice if this was general for instance.

File: doc/source/tutorials/local/Adding_new_data/Select_write_store.rst
Comment (rt17603) [inline doc/source/tutorials/local/Adding_new_data/Select_write_store.rst:1]: Question: is this file being added in addition to the "Adding observation data" tutorial or were they going to be merged in some way? Separate files sound like a good approach - just wondering about avoiding repetition.

File: doc/source/tutorials/local/Adding_new_data/Select_write_store.rst
Comment (rt17603) [inline doc/source/tutorials/local/Adding_new_data/Select_write_store.rst:191]: And search will return details from both object stores? How will retrieve work here? Will it need a store to be passed for instance?

File: openghg/objectstore/_local_store.py
Comment (rt17603) [inline openghg/objectstore/_local_store.py:62]: Missing a doc-string

**Comment by gareth-j** [comment]
And you can add as many stores as you want.

File: openghg/standardise/_surface.py
Comment (rt17603) [inline openghg/standardise/_surface.py:1]: Does this need to be split out from the other standardise functions? And if so, should all the different data types be split out into their own module?

File: openghg/store/_obssurface.py
Comment (rt17603) [inline openghg/store/_obssurface.py:389]: So I guess we should make an explicit note that this is no longer supported? Maybe in the CHANGELOG.md?

**Review by rt17603 — COMMENTED** [review]
I've looked over the changes made so far as much as I can and looks great to me so far!

Main question was around standardise_ inputs and how this is being generalised for the different data types but this may be something you're planning to address but haven't yet.

File: doc/source/tutorials/local/Adding_new_data/Select_write_store.rst
Comment (rt17603) [inline doc/source/tutorials/local/Adding_new_data/Select_write_store.rst:2]: Question: What steps does someone have to take if they want to add a second object store later on? Would they run quickstart again?

**Comment by rt17603** [comment]
@gareth-j - was this PR ready to review then? And did you have a look at the previously added comments or have they been superceded?

File: doc/source/tutorials/local/Adding_new_data/Select_write_store.rst
Comment (gareth-j) [reply doc/source/tutorials/local/Adding_new_data/Select_write_store.rst]: I've removed this file now and hopefully explained how to select a write store in each of the tutorials where it may come up.

File: doc/source/tutorials/local/Adding_new_data/Select_write_store.rst
Comment (gareth-j) [reply doc/source/tutorials/local/Adding_new_data/Select_write_store.rst]: File removed.

Comment (gareth-j) [reply doc/source/tutorials/local/Adding_new_data/Select_write_store.rst]: SearchResults has been updated to be able to retrieve from any store results come from. A single SearchResults object is returned.

File: openghg/dataobjects/_datahandler.py
Comment (gareth-j) [reply openghg/dataobjects/_datahandler.py]: Renamed that.

File: openghg/objectstore/_local_store.py
Comment (gareth-j) [reply openghg/objectstore/_local_store.py:62]: Fixed.

File: openghg/standardise/_standardise.py
Comment (gareth-j) [reply openghg/standardise/_standardise.py:9]: Yeah I thought it'd be tidier than one big file. Store can be passed into each of the functions now.

File: openghg/store/_obssurface.py
Comment (gareth-j) [reply openghg/store/_obssurface.py:389]: Added.

File: openghg/standardise/_surface.py
Comment (gareth-j) [reply openghg/standardise/_surface.py:1]: I think it might be tidier instead of one big file, what do you think?

File: openghg/standardise/_standardise.py
Comment (rt17603) [reply openghg/standardise/_standardise.py:9]: With this cause difficult merge conflicts do you think though? I'm pretty sure I've had to edit all of the standardise functions a few times when updating the way the data is stored so may be tricky before these details have been collated.

File: doc/source/tutorials/local/Adding_new_data/Select_write_store.rst
Comment (gareth-j) [reply doc/source/tutorials/local/Adding_new_data/Select_write_store.rst]: Docs have now been updated.

File: openghg/standardise/_surface.py
Comment (rt17603) [reply openghg/standardise/_surface.py:1]: I'm just thinking this may cause some difficult merge conflicts at the moment with `new_object_store` so if we could hold off on making that kind of change that may be better? I suppose it does depend on how the change has been implemented.

**Comment by gareth-j** [comment]
> @gareth-j - was this PR ready to review then? And did you have a look at the previously added comments or have they been superceded?

Yeah I think it's ready for review now. I've hopefully resolved all the previous comments.

File: openghg/retrieve/ceda/_retrieve.py
Comment (rt17603) [inline openghg/retrieve/ceda/_retrieve.py:232]: What problem is the context manager API solving do you think (e.g. rather than `ObsSurface.store_data(data=to_store, bucket=bucket)`)? Just wondering as it seems to be adding complexity when calling the `ObsSurface.store_data` method (and others) now this is no longer a static method?

File: openghg/store/_metstore.py
Comment (rt17603) [inline openghg/store/_metstore.py:1]: For this file - there was a PR that Elena was working on to try and resurrect this.

File: openghg/store/_rank.py
Comment (rt17603) [inline openghg/store/_rank.py:1]: Is this change related to this PR?

**Comment by rt17603** [comment]
@gareth-j If I try and run the tests locally I'm getting the error:
```
E   ModuleNotFoundError: No module named 'openghg.objectstore._integrity'
```
So looks like there's another module which needs to be commited and pushed for this as well.

File: tests/dataobjects/test_datamanager.py
Comment (rt17603) [inline tests/dataobjects/test_datamanager.py:42]: Can this commented out code be deleted? Or is this still needed?

File: tests/dataobjects/test_datamanager.py
Comment (rt17603) [inline tests/dataobjects/test_datamanager.py:167]: Was this deleted to streamline the test here?

File: tests/retrieve/icos/test_retrieve_icos.py
Comment (rt17603) [inline tests/retrieve/icos/test_retrieve_icos.py:108]: Should we add back the `# update_mismatch=True` for consistency?

File: tests/store/test_metstore.py
Comment (rt17603) [inline tests/store/test_metstore.py:1]: Is this change related to this PR? Similiarly to comment on _metstore.py, we would likely want to be careful this didn't interfere with any changes on PR #615

**Review by rt17603 — CHANGES_REQUESTED** [review]
I've gone through the changes, the tutorials and run the tests and this seems good.

I have included some questions to be addressed below including a discussion of the rationale for the context manager.

The main changes requested here are unrelated to the core functionality being added and mainly around tidying up potential additional changes within this PR.

File: openghg/store/_rank.py
Comment (gareth-j) [reply openghg/store/_rank.py:1]: Hmm that looks like that was collateral damage, I'll see if I can resurrect it.

File: openghg/retrieve/ceda/_retrieve.py
Comment (gareth-j) [reply openghg/retrieve/ceda/_retrieve.py:232]: It has simplified the code in the base class quite a lot and makes opening and closing the metastore easier, removing the need for `load_metadatastore`, `metadatastore.close` etc across the code. It won't have to be used in many places but I think adds quite a bit of safety and ensures things are written correctly.

**Comment by gareth-j** [comment]
> Could you also update the description to include PR #496 and include details that the stated criteria there have been met and if not details of the change (and why). Namely:
> 
> ```
> - Multiple object stores should be treated equally (no "precedence" between them)
> - Add ability to specify which object store to add to, search and to retrieve from
>    - Always have to specify which one to add to for now?
>    - But don't need to explicitly specify which object store to search / retrieve from?
> - Allow searching all object stores (can return multiple SearchResults objects)
> - Allow retrieving from all object stores, as long as there is no ambiguity otherwise fail
> - Allow read/write equally from/to all object stores for now (consider read-only object stores in future or allow this be managed through group assignment and permissions)
> ```

OK done

**Comment by gareth-j** [comment]
@rt17603 due to the reworking of how data is assigned (now done in the base class) I had to rework the `update_metadata` setup. It's now done in `BaseStore.assign_data` (see https://github.com/openghg/openghg/blob/2d81829c28eb5ee657487c2611369b5564a9468c/openghg/store/base/_base.py#L140)

**Comment by rt17603** [comment]
If appropriate, could you also link up the issue #496 so this gets automatically closed (e.g. add "Closes #496" to the PR description).

**Comment by rt17603** [comment]
> @rt17603 due to the reworking of how data is assigned (now done in the base class) I had to rework the `update_metadata` setup. It's now done in `BaseStore.assign_data` (see
> 
> https://github.com/openghg/openghg/blob/2d81829c28eb5ee657487c2611369b5564a9468c/openghg/store/base/_base.py#L140
> )
> 
> All tests pass and hopefully it does what you expect it to.

Was this related to PR #652 ? I did add a test around checking these values were updated within [tests/db_integrity/test_data_updates.py](https://github.com/openghg/openghg/pull/652/files#diff-2072dc8562fd0abf3ac6413a22fa67066dee4ee02141ce4c3d7e7cb59677ca24) - `test_metadata_update` so if that still passes that should be fine.

**Comment by gareth-j** [comment]
> > @rt17603 due to the reworking of how data is assigned (now done in the base class) I had to rework the `update_metadata` setup. It's now done in `BaseStore.assign_data` (see
> > https://github.com/openghg/openghg/blob/2d81829c28eb5ee657487c2611369b5564a9468c/openghg/store/base/_base.py#L140
> > 
> > )
> > All tests pass and hopefully it does what you expect it to.
> 
> @gareth-j Was this related to PR #652 ? I did add a test around checking these values were updated within [tests/db_integrity/test_data_updates.py](https://github.com/openghg/openghg/pull/652/files#diff-2072dc8562fd0abf3ac6413a22fa67066dee4ee02141ce4c3d7e7cb59677ca24) - `test_metadata_update` on that PR so if that still passes that should be fine.

Yeah, those tests pass and we're just making sure that the Datasource and metastore metadata is synchronised.

**Review by rt17603 — APPROVED** [review]

**Comment by rt17603** [comment]
> > > @rt17603 due to the reworking of how data is assigned (now done in the base class) I had to rework the `update_metadata` setup. It's now done in `BaseStore.assign_data` (see
> > > https://github.com/openghg/openghg/blob/2d81829c28eb5ee657487c2611369b5564a9468c/openghg/store/base/_base.py#L140
> > > 
> > > )
> > > All tests pass and hopefully it does what you expect it to.
> > 
> > 
> > @gareth-j Was this related to PR #652 ? I did add a test around checking these values were updated within [tests/db_integrity/test_data_updates.py](https://github.com/openghg/openghg/pull/652/files#diff-2072dc8562fd0abf3ac6413a22fa67066dee4ee02141ce4c3d7e7cb59677ca24) - `test_metadata_update` on that PR so if that still passes that should be fine.
> 
> Yeah, those tests pass and we're just making sure that the Datasource and metastore metadata is synchronised.

@gareth-j - One thing with this that I noticed - has this removed the ability to include specific keys to update using `update_keys`? If I'm understanding the new code correctly this just tries to update everything?

openghg.store.base._base
```
138            # Add the metadata to the metastore and make sure it's up to date with the metadata stored
139            # in the Datasource
140            datasource_metadata = datasource.metadata()
141            if new_ds:
142                self._metastore.insert(datasource_metadata)
143            else:
144                self._metastore.update(datasource_metadata, tinydb.where("uuid") == datasource.uuid())
```

**Review by rt17603 — COMMENTED** [review]
Would like to resolve the conversation around changes to update metadata functionality before merging this.

**Comment by gareth-j** [comment]
> > > > @rt17603 due to the reworking of how data is assigned (now done in the base class) I had to rework the `update_metadata` setup. It's now done in `BaseStore.assign_data` (see
> > > > https://github.com/openghg/openghg/blob/2d81829c28eb5ee657487c2611369b5564a9468c/openghg/store/base/_base.py#L140
> > > > 
> > > > )
> > > > All tests pass and hopefully it does what you expect it to.
> > > 
> > > 
> > > @gareth-j Was this related to PR #652 ? I did add a test around checking these values were updated within [tests/db_integrity/test_data_updates.py](https://github.com/openghg/openghg/pull/652/files#diff-2072dc8562fd0abf3ac6413a22fa67066dee4ee02141ce4c3d7e7cb59677ca24) - `test_metadata_update` on that PR so if that still passes that should be fine.
> > 
> > 
> > Yeah, those tests pass and we're just making sure that the Datasource and metastore metadata is synchronised.
> 
> @gareth-j - One thing with this that I noticed - has this removed the ability to include specific keys to update using `update_keys`? If I'm understanding the new code correctly this just tries to update everything?
> 
> openghg.store.base._base
> 
> ```
> 138            # Add the metadata to the metastore and make sure it's up to date with the metadata stored
> 139            # in the Datasource
> 140            datasource_metadata = datasource.metadata()
> 141            if new_ds:
> 142                self._metastore.insert(datasource_metadata)
> 143            else:
> 144                self._metastore.update(datasource_metadata, tinydb.where("uuid") == datasource.uuid())
> ```

Yeah now the metadata is just completely in sync between the metastore and the Datasource, do we want it to work differently?

**Comment by rt17603** [comment]
> Yeah now the metadata is just completely in sync between the metastore and the Datasource, do we want it to work differently?

I think that should probably be fine - I feel like there were some internal details on Datasource that we weren't copying across but if I can't immediately identify these then this shouldn't stop this change going in.

File: openghg/store/base/_base.py
Comment (rt17603) [inline openghg/store/base/_base.py:43]: Is this a question to be addressed later? May want to add a TODO at the start if so?

File: tests/store/test_obssurface.py
Comment (gareth-j) [reply tests/store/test_obssurface.py:21]: Yeah, this is pretty much just checking the Python implementation so might not be best way but I just wanted to make sure the way I setup the handling of errors worked as I expected.

File: openghg/store/_eulerian_model.py
Comment (rt17603) [inline openghg/store/_eulerian_model.py:40]: Two things:
1. I believe `logger.exception` is mainly called from within a try-except statement from looking at - https://docs.python.org/3/library/logging.html#logging.Logger.exception which says "This method should only be called from an exception handler.". So I think this should be [`logger.error()` instead](https://docs.python.org/3/library/logging.html#logging.Logger.error)
2. Did you want this to raise the error or just log it?

File: tests/store/test_obssurface.py
Comment (rt17603) [reply tests/store/test_obssurface.py:21]: So is this checking that raising a `ValueError` from in the `with` statement is doing the right thing?

File: openghg/store/_eulerian_model.py
Comment (gareth-j) [reply openghg/store/_eulerian_model.py:40]: Ohh OK, I must have got them mixed up. I'll change it to be logger.error now. I want the error to propagate.

File: tests/store/test_obssurface.py
Comment (gareth-j) [reply tests/store/test_obssurface.py:21]: Yeah, in case people do something else with `obs` that raises an error. Then we test what happens if `assign_data` raises an error below.

Comment (rt17603) [reply tests/store/test_obssurface.py:21]: I think a quick comment here to that effect would help make sure we're clear on what this is checking for. I imagine we could get a bit confused otherwise.

File: openghg/store/_eulerian_model.py
Comment (gareth-j) [reply openghg/store/_eulerian_model.py]: OK done

**Review by rt17603 — COMMENTED** [review]
A few quick comments/questions:
 - Does the change to the `__exit__` check need to be added to all of the top level classes or could this just be included on the BaseStore class and be inherited?
 - If it does need to be included on all the top level classes, this should also be updated within `_boundary_conditions.py`

File: openghg/store/base/_base.py
Comment (gareth-j) [reply openghg/store/base/_base.py]: Ah I started off with that but forgot to remove it. I've deleted it now.

**Review by rt17603 — APPROVED** [review]

**Pull request merged**
